src/framework/matrix.hpp

```diff
@@ -279,6 +142,9 @@ class matrix {
   // Addressing elements by matrix representation
   T &operator()(size_t row, size_t col);
   T operator()(size_t row, size_t col) const;
+  // Addressing elements by row or column
+  std::vector<T> row_index(size_t row) const;
```

Comment:
I guess you are using these functions later (because I can't see them in this PR).

Reply:
I need to update the tests - notice that the function signature of eigensystem_psd was changed from:
```
void eigensystem_psd(const matrix<std::complex<float_t>>& psd_matrix,
                /* out */ std::vector<std::complex<float_t>>& eigenvalues,
                /* out */ std::vector<std::vector<std::complex<float_t>>>& eigenvectors){
```
to
```
void eigensystem_psd(const matrix<std::complex<float_t>>& psd_matrix,
               /* out */ std::vector<T>& eigenvalues,
                /* out */ matrix<std::complex<T>>& eigenvectors) {}
```
This + adding the matrix getCol/Row methods allow me to avoid a copy within the eigensystem_psd function while still allowing the user to pull out the eigenvectors (stored in each row of the eigenvector matrix).
Granted, the copy is not skipped completely but is just moved to when the eigenvector is pulled out (on getRow/Col) as you noticed below...

Reply:
also @chriseclectic - can we confirm that the eigenvalues here (for PSD hermitian) are ALWAYS real / NEVER complex ...? if this isn't true I feel like the lapack function signatures aren't correct..? They all have the eigenvalue output types being arrays of float/double rather than array of complex<float>

Reply:
They are all real all the time - if we need to cast to complex, that should be on the user - it's a really just an extra allocation and copy that isn't necessary given the assumed properties of the input - if the eigenvalues end up complex, the input constraints must not have been obeyed and an error should be/is raised.